File: 230.java
// Leetcode 230 Kth Smallest Element in a BST
// References:https://leetcode.com/problems/kth-smallest-element-in-a-bst/solutions/63783/two-easiest-in-order-traverse-java/?envType=study-plan-v2&envId=top-interview-150

/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public int kthSmallest(TreeNode root, int k) {
        Stack<TreeNode> stack = new Stack<TreeNode>();
        TreeNode p = root;
        int count = 0;

        while (p != null || !stack.isEmpty()){
            if (p != null) {
                stack.push(p); // similar to calling dfs(root.left)
                p = p.left;
            } else {
                p = stack.pop();

                //---------- real code starts here ----------
                count++;
                if (count == k){
                    return p.val;
                }
                //-------------------------------------------
                p = p.right;  // similar to calling dfs(root.right)
            }
        }
        return Integer.MIN_VALUE;
    }
}

File: 230.js
// Leetcode 230 Kth Smallest Element in a BST
// References: https://leetcode.com/problems/kth-smallest-element-in-a-bst/solutions/1095674/js-recursive-iterative-inorder-solutions/?envType=study-plan-v2&envId=top-interview-150

/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @param {number} k
 * @return {number}
 */
var kthSmallest = function(root, k) {
    const stack = []
    let count = 1
    let node = root

    while (node || stack.length){
        while (node){
            stack.push(node)
            node = node.left
        }
        node = stack.pop()
        if (count === k) return node.val
        else count++;
        node = node.right;
    }

};

// Recursive
const kthSmalles2 = (root, k) => {
    let n = 0;
    let res;
    const inorder = (root) => {
        if (!root) return;
        inorder(root.left);
        if (n++ < k) res = root.val;
        inorder(root.right);
    };
    inorder(root);
    return res;
}